Milestone 2: API / glue library docs (#26)
mlabs-haskell/hydra-auction-offchain: docs/API.md

# Hydra Auction Frontend API

This document describes the API of the frontend glue library 
that allows the Hydra Auction frontend to execute auction-related operations (also known as "contracts"). 
The library is based on CTL. 

## Executing operations

The library can be used by importing a set of types and functions from `hydra-auction-offchain` package 
as shown on a test page in the `/demo` folder.

Operations are exposed as asynchronous functions (eponymous to names in this document) 

> that yield a value of type `ContractOutput` (see below).

Reviewer: That's not exactly true for all exposed operations, since some of them (such as `queryAuctions` or `discoverBidders`) do not define typed errors and therefore do not return `ContractOutput`.

```suggestion
and most of them yield a value of type `ContractOutput` (see below).
```

      <td></td>
    </tr>
    <tr>
      <td align="left">discoverSellerSignature</td>
      <td></td>
      <td></td>
      <td></td>
      <td colspan="2">+</td>
      <td></td>
    </tr>
    <tr>
      <td align="left">placeBid</td>
      <td></td>
      <td></td>
      <td></td>
      <td colspan="2">+</td>
      <td></td>
    </tr>
    <tr>
      <td align="left">claimAuctionLotBidder</td>
      <td></td>
      <td></td>
      <td></td>
      <td>+</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td align="left">claimDepositLoser</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td>+</td>
      <td></td>
    </tr>
    <tr>
      <td align="left">announceDelegateGroup</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td>+</td>
    </tr>
  </tbody>
</table>

## Operations available for all users

### queryAuctions

Query all currently existing auctions discoverable on-chain,
filtering out invalid ones (i.e. having invalid auction terms).

Takes no parameters.

Returns the list of `AuctionInfo` values, which consists of `AuctionTerms` 
plus some on-chain auction parameters, 

> like auction the token currency symbol and addresses of auction validators.

Reviewer: ```suggestion
like auction currency symbol (auction ID) and addresses of auction validators.
```

### cleanupAuction

The cleanup procedure is supposed to happen after the purchase deadline,

> burning down all auction-related assets and thus removing all information about the auction from the blockchain.

Reviewer: I think we should also mention that during cleanup all auction-related leftover utxos are consumed, like auction escrow utxo and auction metadata utxo.

Takes `AuctionInfo` as an argument.

Returns the hash of a submitted cleaning transaction.

### queryStandingBidState

The one and only (but potentially absent) standing bid for an auction can be obtained using this operation.

Takes `AuctionInfo` as an argument.

> Returns optionally missing `Maybe BidTerm` value.

Reviewer: ```suggestion
Returns optionally missing `Maybe BidTerms` value.
```

## Available to Sellers

### announceAuction

Announce an auction on-chain by minting tokens,
saving the auction metadata on-chain,
and placing the auction lot into the auction's escrow.

Takes:
  1. Auction terms;
  2. Optional additional UTxOs that allow the seller to cover auction lot `Value` in some cases. This can be useful if some portion of the lot is, for example, locked at a multi-signature address.

> Returns `ActionInfo` along with the hash of an announcing transaction.

Reviewer: ```suggestion
Returns `AuctionInfo` along with the hash of an announcing transaction.
```

> This operation can error see type `AnnounceAuctionContractError`.

Reviewer: ```suggestion
This operation can error (see type `AnnounceAuctionContractError`).
```
